### ISSUE TITLE
[tune] How to write a correct setting to conduct hyperparam search experiment appropriately?

### ISSUE THREAD

**Issue by guoxuxu** [ray-project/ray#9815]
### What is your question?

1. I noticed that even if I fix the random seed, the best config and best metric found can vary a lot for the same python file executed. So I'm wondering <u> what's the correct and professional way to make sure my hyperparameter search experiment is running towards a potential best config and the metric result?

I tried to figure out this question by specifying ``` stop``` and ```num_samples``` in ```tune.run()```. Specifically, I guess (1) by setting ```stop{metric: 0.96, "training_iteration": 10}``` the larger the ```training_iteration```, the better the results; (2) the larger the ```num_samples``` is, the more possible the experiment is approaching the best config. I guess only (2) can encourage the experiment running towards the best result? If so, usually how large can we set ```num_samples```?  Please also let me know if other attributes in tune can help.

 It seems that the maximum iterations allowed for each trial are 100? ```training_iteration``` can only be effective when it is set to some value lower than 100? as shown below

![image](https://user-images.githubusercontent.com/29363464/88908335-4f8b3900-d28c-11ea-89de-ff4a4e5bc89b.png)
![image](https://user-images.githubusercontent.com/29363464/88908354-55811a00-d28c-11ea-9b30-b8356930114f.png)

My code for ray tune initialization:
```
    if args.address:
        ray.init(address=args.address)
    else:
        ray.init(num_cpus=os.cpu_count(), num_gpus=torch.cuda.device_count(), temp_dir=temp_dir)
    sched = AsyncHyperBandScheduler(time_attr="training_iteration", metric="mean_accuracy")

    space = {
        "lr": hp.loguniform("lr", -10, -1),  
        "momentum": hp.uniform("momentum", 0.1, 0.9), 
    }
    metric = "mean_accuracy"
    algo = HyperOptSearch(
        space=space,
        max_concurrent=8, 
        metric=metric,
        mode="max"
    )

    analysis = tune.run(
        train_mnist,
        name="exp",
        verbose=1, 
        scheduler=sched,
        search_alg=algo,
        stop={
            metric: 0.95,
            "training_iteration": 10
        },
        resources_per_trial={
            "cpu": 1,
            "gpu": 0.1
        },
        num_samples=20,
        config={
            "use_gpu": True
        })
```

2. To wrap the whole experiment to conduct a complete hyperparameter search, is it better to write a Trainable Class ```class train_mnist(tune.Trainable)``` than a Trainable function ```def train_mnist(config)```? What's their true difference? Through ```nvidia-smi```, I saw that a trainable function shows ```ray::WrappedTrackFunc.train()``` in version 0.8.5 and ```ray::ImplicitFunc.train()``` in version 0.8.6 while a trainable class shows ```ray::Train.__init__()``` in version 0.8.5. what do they imply?

*Ray version and other system information (Python version, TensorFlow version, OS):*
Python 3.7.4
OS: linux
Ray: 0.8.5

**Comment by richardliaw** [comment]
1. **Seeding**: set the seed before tune.run is called AND inside your trainable class/function
2. Trainable function is preferable; we're slowly phasing out the class abstraction; it is only there for

**Comment by guoxuxu** [comment]
Thank you very much! ^^
